File: skills/code-architect/references/code-examples/cpp/Mediator.cpp
// Mediator — centralize communication between components.
#include <iostream>
#include <string>
#include <vector>

class Mediator;

// Colleague base knows only the mediator.
class Participant {
protected:
    Mediator* mediator = nullptr;
    std::string name;
public:
    explicit Participant(std::string n) : name(std::move(n)) {}
    virtual ~Participant() = default;
    void setMediator(Mediator* m) { mediator = m; }
    const std::string& id() const { return name; }
    virtual void receive(const std::string& from, const std::string& msg) {
        std::cout << name << " got from " << from << ": " << msg << "\n";
    }
    void send(const std::string& msg);
};

class Mediator {
    std::vector<Participant*> members;
public:
    void registerMember(Participant* p) {
        members.push_back(p);
        p->setMediator(this);
    }
    void broadcast(Participant* sender, const std::string& msg) {
        for (auto* m : members)
            if (m != sender) m->receive(sender->id(), msg);
    }
};

void Participant::send(const std::string& msg) {
    if (mediator) mediator->broadcast(this, msg);
}

int main() {
    Mediator hub;
    Participant a{"A"}, b{"B"}, c{"C"};
    hub.registerMember(&a);
    hub.registerMember(&b);
    hub.registerMember(&c);

    a.send("hello everyone");
    return 0;
}

File: skills/code-architect/references/code-examples/ruby/mediator.rb
# frozen_string_literal: true

# Mediator — centralize communication between components.

# Mediator coordinates colleagues so they don't reference each other directly.
class ChatRoom
  def initialize = @members = {}

  def join(participant)
    @members[participant.name] = participant
    participant.room = self
  end

  # Colleagues talk through the mediator only.
  def relay(from:, text:)
    @members.each_value do |member|
      member.receive(from: from, text: text) unless member.name == from
    end
  end
end

# Colleague knows the mediator but not its peers.
class Participant
  attr_reader :name
  attr_accessor :room

  def initialize(name) = @name = name

  def send(text)
    room.relay(from: @name, text: text)
  end

  def receive(from:, text:)
    puts "#{@name} got from #{from}: #{text}"
  end
end

if __FILE__ == $PROGRAM_NAME
  room = ChatRoom.new
  ana = Participant.new("ana")
  ben = Participant.new("ben")
  cid = Participant.new("cid")
  [ana, ben, cid].each { |p| room.join(p) }

  ana.send("hi all")
end
